### PR TITLE
Fix for HADOOP-12169

### PR DESCRIPTION
**Pull request by kadaan**
Added a fix for a hadoop issue (https://issues.apache.org/jira/browse/HADOOP-12169) which affects the s3a filesystem and results in duplicate files appearing in the results of ListStatus.  In the process, extracted a base class for all FsHelper classes based on the hadoop filesystem.

### CODE REVIEW

File: gobblin-core/src/main/java/gobblin/source/extractor/hadoop/HadoopFsHelper.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
+package gobblin.source.extractor.hadoop;
```
Comment (liyinan926) [inline gobblin-core/src/main/java/gobblin/source/extractor/hadoop/HadoopFsHelper.java]: Missing open source header.

File: gobblin-core/src/main/java/gobblin/source/extractor/hadoop/HadoopFsHelper.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+import gobblin.configuration.ConfigurationKeys;
```
Comment (liyinan926) [inline gobblin-core/src/main/java/gobblin/source/extractor/hadoop/HadoopFsHelper.java]: Wrong order of imports, see https://github.com/linkedin/gobblin/wiki/CodingStyle.

File: gobblin-core/src/main/java/gobblin/source/extractor/hadoop/HadoopFsHelper.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.List;
+
+public abstract class HadoopFsHelper implements FileBasedHelper {
```
Comment (liyinan926) [inline gobblin-core/src/main/java/gobblin/source/extractor/hadoop/HadoopFsHelper.java]: No Javadoc.

File: gobblin-core/src/main/java/gobblin/source/extractor/hadoop/HadoopFsHelper.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+  public FileSystem getFileSystem() {
+    return this.fs;
+  }
+
+  @Override
+  public void connect() throws FileBasedHelperException {
+    String uri = state.getProp(ConfigurationKeys.SOURCE_FILEBASED_FS_URI);
+    try {
+      if (uri == null) {
```
Comment (liyinan926) [inline gobblin-core/src/main/java/gobblin/source/extractor/hadoop/HadoopFsHelper.java]: Please use `Strings.isNullOrEmpty(uri)` instead.

File: gobblin-core/src/main/java/gobblin/source/extractor/hadoop/HadoopFsHelper.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+  protected HadoopFsHelper(State state, Configuration configuration) {
+    this.state = state;
+    this.configuration = configuration;
+  }
+
+  protected State getState() {
+    return this.state;
+  }
+
+  public FileSystem getFileSystem() {
```
Comment (liyinan926) [inline gobblin-core/src/main/java/gobblin/source/extractor/hadoop/HadoopFsHelper.java]: Does this need to be `public`?

File: gobblin-core/src/main/java/gobblin/source/extractor/hadoop/HadoopFsHelper.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    } catch (URISyntaxException e) {
+      throw new FileBasedHelperException("Malformed uri " + uri + " due to " + e.getMessage(), e);
+    } catch (InterruptedException e) {
+      throw new FileBasedHelperException("Interrupted exception is caught when getting the proxy file system", e);
+    }
+  }
+
+  @Override
+  public List<String> ls(String path) throws FileBasedHelperException {
+      List<String> results = new ArrayList<>();
```
Comment (liyinan926) [inline gobblin-core/src/main/java/gobblin/source/extractor/hadoop/HadoopFsHelper.java]: You can use an `ImmutableList.Builder` and return an `ImmutableList`.

File: gobblin-core/src/main/java/gobblin/source/extractor/hadoop/AvroFsHelper.java [hunk 7]
```diff
@@ -98,37 +52,13 @@ public void close() throws FileBasedHelperException {
...
-      }
-    }
-  }
-
-  @Override
   public InputStream getFileStream(String path) throws FileBasedHelperException {
     try {
       Path p = new Path(path);
-      InputStream in = this.fs.open(p);
+      InputStream in = this.getFileSystem().open(p);
```
Comment (liyinan926) [inline gobblin-core/src/main/java/gobblin/source/extractor/hadoop/AvroFsHelper.java]: Please put this in `try (InputStream in = this.getFileSystem().open(new Path(path)))`.

File: gobblin-core/src/main/java/gobblin/source/extractor/hadoop/AvroFsHelper.java [hunk 8]
```diff
@@ -139,14 +69,14 @@ public InputStream getFileStream(String path) throws FileBasedHelperException {
   public Schema getAvroSchema(String file) throws FileBasedHelperException {
     DataFileReader<GenericRecord> dfr = null;
     try {
-      if (state.getPropAsBoolean(ConfigurationKeys.SHOULD_FS_PROXY_AS_USER,
+      if (this.getState().getPropAsBoolean(ConfigurationKeys.SHOULD_FS_PROXY_AS_USER,
           ConfigurationKeys.DEFAULT_SHOULD_FS_PROXY_AS_USER)) {
         dfr =
-            new DataFileReader<GenericRecord>(new ProxyFsInput(new Path(file), this.fs),
```
Comment (liyinan926) [inline gobblin-core/src/main/java/gobblin/source/extractor/hadoop/AvroFsHelper.java]: Please use a `Closer` to register `dfr` and `FsInput`/`ProxyFsInput`. And please rename `dfs` to some more meaningful name.

File: gobblin-core/src/main/java/gobblin/source/extractor/hadoop/AvroFsHelper.java [hunk 9]
```diff
@@ -165,16 +95,16 @@ public Schema getAvroSchema(String file) throws FileBasedHelperException {
 
   public DataFileReader<GenericRecord> getAvroFile(String file) throws FileBasedHelperException {
```
Comment (liyinan926) [inline gobblin-core/src/main/java/gobblin/source/extractor/hadoop/AvroFsHelper.java]: Please add Javadoc for this method and make it clear that the caller is responsible for closing the returned `DataFileReader`.

File: gobblin-core/src/main/java/gobblin/source/extractor/hadoop/HadoopFsHelper.java (hunk 5 above)
Comment (kadaan) [inline gobblin-core/src/main/java/gobblin/source/extractor/hadoop/HadoopFsHelper.java]: `HadoopFsHelper` needs to be public because people implementing other FsHelper outside of the gobblin project will need access.  `getFileSystem` needs to be public because it is used by `SizeAwareFileBasedHelperDecorator`.

Comment (liyinan926) [inline gobblin-core/src/main/java/gobblin/source/extractor/hadoop/HadoopFsHelper.java]: Got it. I meant `getFileFileSystem`.

File: gobblin-core/src/main/java/gobblin/source/extractor/hadoop/HadoopFsHelper.java (hunk 6 above)
Comment (kadaan) [inline gobblin-core/src/main/java/gobblin/source/extractor/hadoop/HadoopFsHelper.java]: I would prefer to not make this change at the moment.  This code was pulled directly from `AvroFsHelper.java`.  This brings up the broader question as to when changes like this should be made.  My preference would be to minimize the changes, only fixing obvious issues.  If another PR is needed, to address these issues, then so be it.

Comment (liyinan926) [inline gobblin-core/src/main/java/gobblin/source/extractor/hadoop/HadoopFsHelper.java]: Fine.

File: gobblin-core/src/main/java/gobblin/source/extractor/hadoop/AvroFsHelper.java (hunk 7 above)
Comment (kadaan) [inline gobblin-core/src/main/java/gobblin/source/extractor/hadoop/AvroFsHelper.java]: I don't think this can be put in a try-with-resources statement because the `InputStream` is potentially returned on line 63.

Comment (liyinan926) [inline gobblin-core/src/main/java/gobblin/source/extractor/hadoop/AvroFsHelper.java]: Ah. Didn't realize that. Then please make it clear in the Javadoc that it's the caller's responsibility to close it.

File: gobblin-core/src/main/java/gobblin/source/extractor/hadoop/AvroFsHelper.java (hunk 8 above)
Comment (kadaan) [inline gobblin-core/src/main/java/gobblin/source/extractor/hadoop/AvroFsHelper.java]: I'm going to pass on this as well.  This is legacy code and a refactoring that I'm not prepared to test right now.

File: gobblin-core/src/main/java/gobblin/source/extractor/hadoop/HadoopFsHelper.java (hunk 1 above)
Comment (kadaan) [inline gobblin-core/src/main/java/gobblin/source/extractor/hadoop/HadoopFsHelper.java]: Added

File: gobblin-core/src/main/java/gobblin/source/extractor/hadoop/AvroFsHelper.java (hunk 8 above)
Comment (liyinan926) [inline gobblin-core/src/main/java/gobblin/source/extractor/hadoop/AvroFsHelper.java]: Yes, this is legacy code but the way it uses the `Closeable`s is not done right and risks leaking resources, so this must be fixed.

Comment (kadaan) [inline gobblin-core/src/main/java/gobblin/source/extractor/hadoop/AvroFsHelper.java]: This will be done in a different PR.

Comment (liyinan926) [inline gobblin-core/src/main/java/gobblin/source/extractor/hadoop/AvroFsHelper.java]: OK.

**Comment by liyinan926** [comment]
LGTM.
